sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubPartitionHashJoin.scala

```diff
@@ -0,0 +1,890 @@
+    if (!hasNext) throw new NoSuchElementException()
+    // Get the next partition ids for this output. It will try to coalesce small
+    // partitions to reach "avgNumBatches" batches as much as possible.
+    val (ids, poss) = nextPartitionIdsAndPoss().unzip
+    // Take over the batches of one or multiple partitions according to the ids. And
+    // concatenate them in a single batch.
+    val spillBatches = closeOnExcept(ArrayBuffer.empty[SpillableColumnarBatch]) { buf =>
+      ids.foreach { id =>
+        buf ++= batchSubPartitioner.releaseBatchesByPartition(id)
+      }
+      buf
+    }
+    val retBatch = GpuSubPartitionHashJoin.concatSpillBatchesAndClose(
+      spillBatches, spillCallback)
+
+    // Update the remaining partitions
+    poss.foreach(remainingPartIdsAndSizes.remove)
+    (ids, retBatch)
+  }
+
+  private[this] def nextPartitionIdsAndPoss(): Array[(Int, Int)] = {
+    val ret = ArrayBuffer.empty[(Int, Int)]
+    var totalBatches = 0
+    // always append the first one
+    val (firstId, firstBatchCnt) = remainingPartIdsAndSizes.head
+    ret += ((firstId, 0))
+    totalBatches += firstBatchCnt
+    // For each output, try to collect small nonempty partitions to reach "avgNumBatches"
+    // batches as much as possible.
+    // e.g. We have 5 partitions as below. The first partition has two batches, and the
+    // other two partitions have one batch for each.
+    //    p1: (3,4)
+    //    p2: (1,2), (11, 22)
+    //    p3: (5,6)
+    //    p4: <empty>
+    //    p5: <empty>
+    // Then the p1 and p3 will be put together in one output.
+    //    next: (3,4), (5,6)
+    //    next: (1,2), (11, 22)
+    //    next: <empty>
+    //    next: <empty>
+    if (firstBatchCnt > 0) {
+      remainingPartIdsAndSizes.tail.zipWithIndex.foreach { case ((id, batchCnt), pos) =>
+        // Do not coalesce empty partitions, and output them one by one.
+        if (batchCnt > 0 && (totalBatches + batchCnt) <= avgNumBatches) {
+          ret += ((id, pos))
+          totalBatches += batchCnt
+        }
+      }
+    }
+    ret.toArray
+  }
+}
+
+/**
+ * An utils class that will take over the two input batches.
+ */
+class BatchPair(
+    private var build: Option[ColumnarBatch],
+    private var stream: Option[ColumnarBatch]) extends AutoCloseable {
+
+  /**
+   * Whether the batch pair is empty.
+   * A pair is empty when both left and right batches are empty.
+   */
+  def isEmpty: Boolean = build.isEmpty && stream.isEmpty
+
+  /**
+   * Get the two batches as a Tuple(build, stream).
+   */
+  def get: (Option[ColumnarBatch], Option[ColumnarBatch]) = {
+    (build, stream)
+  }
+
+  /**
+   * Release the two batches as a Tuple(build, stream).
+   */
+  def release: (Option[ColumnarBatch], Option[ColumnarBatch]) = {
+    val ret = (build, stream)
+    build = None
+    stream = None
+    ret
+  }
+
+  override def close(): Unit = {
+    build.foreach(_.safeClose(new Exception))
```

Comment:
why is `new Exception` is passed here?

Reply:
To let this `safeClose ` to be a true `safeClose `.
See its code below
```
    def safeClose(e: Throwable = null): Unit = {
      if (autoCloseable != null) {
        if (e != null) {
          try {
            autoCloseable.close()
          } catch {
            case suppressed: Throwable => e.addSuppressed(suppressed)
          }
        } else {
          autoCloseable.close()
        }
      }
    }
```